communication/src/AsyncTlsConnection.cpp

```diff
@@ -0,0 +1,390 @@
+    }
+  });
+}
+
+uint32_t AsyncTlsConnection::getReadMsgSize() {
+  // We send in network byte order
+  return ntohl(*reinterpret_cast<uint32_t*>(read_size_buf_.data()));
+}
+
+void AsyncTlsConnection::dispose() {
+  ConcordAssert(!disposed_);
+  LOG_ERROR(logger_, "Closing connection to node " << peer_id_.value());
+  disposed_ = true;
+  boost::system::error_code _;
+  read_timer_.cancel(_);
+  write_timer_.cancel(_);
+  tlsTcpImpl_.closeConnection(peer_id_.value());
+}
+
+void AsyncTlsConnection::send(std::vector<char>&& raw_msg) {
+  std::lock_guard<std::mutex> guard(write_lock_);
+  if (queued_size_in_bytes_ + raw_msg.size() > MAX_QUEUE_SIZE_IN_BYTES) {
+    LOG_WARN(logger_, "Outgoing Queue is full. Dropping message with size: " << raw_msg.size());
+    return;
+  }
+
+  queued_size_in_bytes_ += raw_msg.size();
+  out_queue_.emplace_back(OutgoingMsg(std::move(raw_msg)));
+  if (out_queue_.size() == 1) {
+    write();
+  }
+}
+
+// Invariant:  `write_lock_` is held when this function is called
+void AsyncTlsConnection::write() {
+  while (!out_queue_.empty() &&
+         out_queue_.front().send_time + STALE_MESSAGE_TIMEOUT < std::chrono::steady_clock::now()) {
+    auto diff = std::chrono::steady_clock::now() - out_queue_.front().send_time;
+    LOG_WARN(logger_,
+             "Message queued for peer " << peer_id_.value() << " for "
+                                        << std::chrono::duration_cast<std::chrono::seconds>(diff).count()
+                                        << " seconds, with size: " << out_queue_.front().msg.size()
+                                        << " dropped. Message is stale: Exceeded threshold of "
+                                        << STALE_MESSAGE_TIMEOUT.count() << " seconds.");
+    queued_size_in_bytes_ -= out_queue_.front().msg.size();
+    out_queue_.pop_front();
+  }
+  if (out_queue_.empty()) {
+    return;
+  }
+
+  auto self = shared_from_this();
+  startWriteTimer();
+  boost::asio::async_write(
+      *socket_, boost::asio::buffer(out_queue_.front().msg), [this, self](const boost::system::error_code& ec, auto _) {
+        if (ec) {
+          if (ec == boost::asio::error::operation_aborted || disposed_) {
+            // The socket has already been cleaned up and any references are invalid. Just return.
+            return;
+          }
+          LOG_WARN(logger_,
+                   "Write failed to node " << peer_id_.value() << " for message with size "
+                                           << out_queue_.front().msg.size() << ": " << ec.message());
+          return dispose();
+        }
+        // The write succeeded.
+        write_timer_.cancel();
+        std::lock_guard<std::mutex> guard(write_lock_);
+        queued_size_in_bytes_ -= out_queue_.front().msg.size();
+        out_queue_.pop_front();
+        if (!out_queue_.empty()) {
+          write();
+        }
+      });
+}
+
+void AsyncTlsConnection::createSSLSocket(boost::asio::ip::tcp::socket&& socket) {
```

Comment:
Probably just a copy and paste but the coding standard wants us to use `Ssl` instead.
"[...] prefer to capitalize abbreviations as single words, e.g., StartRpc() rather than StartRPC()"

Reply:
going to leave this one for now.